File: go/design-a-stack-with-increment-operation.go
type CustomStack struct {
	MaxSize    int
	Stack      []int
	Increments []int
}

func Constructor(maxSize int) CustomStack {
	// Initialize CustomStack object
	customStack := CustomStack{
		MaxSize:    maxSize,
		Stack:      make([]int, 0),
		Increments: make([]int, 0),
	}
	return customStack
}

func (this *CustomStack) Push(x int) {
	if len(this.Stack) < this.MaxSize {
		// Add element to the stack
		this.Stack = append(this.Stack, x)
		// Add an increment of 0 to the increment's stack
		this.Increments = append(this.Increments, 0)
	}
}

func (this *CustomStack) Pop() int {
	n := len(this.Stack)
	if 0 < n {
		// Pop increment, rollover to the lower level
		increment := this.Increments[n-1]
		this.Increments = this.Increments[:n-1]
		this.Increment(len(this.Increments), increment)
		// Pop element from stack
		peak := this.Stack[n-1]
		this.Stack = this.Stack[:n-1]
		// Return the stack value and the increment
		return peak + increment
	} else {
		return -1
	}
}

func (this *CustomStack) Increment(k int, val int) {
	n := len(this.Increments)
	if 0 < n {
		// Pick smallest from Increment stack size and k
		minimal_index := k
		if n < minimal_index {
			minimal_index = n
		}
		// Update increment in the position of the stack
		this.Increments[minimal_index-1] += val
	}
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * obj := Constructor(maxSize);
 * obj.Push(x);
 * param_2 := obj.Pop();
 * obj.Increment(k,val);
 */

File: python/design-a-stack-with-increment-operation.py
class CustomStack:

    def __init__(self, maxSize: int):
        self.max_size = maxSize
        self.stack = []
        self.increments = []

    def push(self, x: int) -> None:
        if len(self.stack) < self.max_size:
            self.stack.append(x)
            self.increments.append(0)

    def pop(self) -> int:
        if len(self.stack) == 0:
            return -1
        increment = self.increments.pop()
        self.increment(len(self.increments), increment)
        return self.stack.pop() + increment

    def increment(self, k: int, val: int) -> None:
        if 0 < len(self.increments):
            increment_index = min(k, len(self.increments)) - 1
            self.increments[increment_index] += val

